laserstudio/widgets/toolbars/focustoolbar.py

```diff
@@ -0,0 +1,246 @@
+            self.stage,
+            FocusSearchSettings(4000 / objective, 50, 4, False),
+            FocusSearchSettings(200 / objective, 20, 16, False),
+            None,
+        )
+        self.focus_thread = t
+        t.finished.connect(self.magic_focus_finished)
+        t.start()
+
+    def magic_focus_finished(self):
+        """Called when focus search thread has finished."""
+        assert self.focus_thread is not None
+        self.focus_thread = None
+        self.button_magic_focus.setEnabled(True)
+
+    def register(self):
+        """
+        Registers a new focus point. If three focus points are already defined, the
+        farther point is replaced.
+        """
+        pos = self.stage.position
+        if len(self.autofocus_helper) == 3:
+            dists = [
+                np.linalg.norm((Vector(*p).xy - pos.xy).data)
+                for p in self.autofocus_helper.registered_points
+            ]
+            del self.autofocus_helper.registered_points[dists.index(min(dists))]
+        self.autofocus_helper.register(pos.x, pos.y, pos.z)
+
+    def autofocus(self):
+        """
+        Calculate focus for the given position and apply it, if possible.
+        """
+        pos = self.stage.position
+        z = self.autofocus_helper.focus(pos.x, pos.y)
+        if abs(z - pos.z < 250):
+            print("DIFF", z, pos.z)
+            self.stage.position = Vector(pos.x, pos.y, z)
+        else:
+            print("Warning: too big Z difference")
+
+
+class FocusSearchSettings:
+    """Parameters for the focus search procedure."""
+
+    def __init__(self, span: float, steps: int, avg: int, multi_peaks: bool):
+        """
+        :param span: Z search span, in micrometers. Maximum allowed value is
+            1000 µm, for safety purpose. Search will occur in the range
+            [z - span / 2, z + span / 2].
+        :param steps: Number of search steps. Must be greater or equal to 2.
+        :param avg: Image averaging setting. Must be greater or equal to 1.
+        :multi_peaks: If True, peak detection is performed, and the peak with
+            the higher Z value is considered as the correct focus. This is used
+            to distinguish the silicon transistors from the silicon surface. If
+            False, the position with the highest image standard deviation is
+            kept.
+        """
+        assert span > 0
+        assert span < 1000  # In case of bug, prevent large span
+        assert steps >= 2
+        assert avg >= 1
+        self.span = span
+        self.steps = steps
+        self.avg = avg
+        self.multi_peaks = multi_peaks
+
+
+class FocusThread(QThread):
+    """
+    Thread to perform best focus position research by moving a stage and analysing the
+    images of a camera.
+    """
+
+    def __init__(
+        self,
+        camera: CameraNITInstrument,
+        stage,
+        coarse: FocusSearchSettings,
+        fine: FocusSearchSettings | None = None,
```

Comment:
LaserStudio is intended to run on Python 3.9.
"|" is introduced in later version.
Could you use `Optional[FocusSearchSettings]` instead